earth_enterprise/src/server/mod_fdb/portableservice_unittest.cpp

```diff
@@ -0,0 +1,228 @@
+
+class PortableServiceTest : public testing::Test {
+  protected:
+    PortableService service;
+    request_rec rec;
+    request_rec *r = &rec;
+    ArgMap args;
+    PortableServiceTest() {
+      service.RegisterPortable(r, TARGET, GLOBE);
+      write_data.clear();
+    }
+    void TestBadRequest(
+      request_rec* r,
+      const std::string& source_cmd_or_path,
+      const std::string& target_path,
+      const std::string& suffix,
+      int layer_id,
+      bool is_balloon_request,
+      ArgMap& arg_map,
+      const std::string& no_value_arg,
+      fusion_portableglobe::CutSpec* cutspec) {
+        TestRequest(
+          r,
+          source_cmd_or_path,
+          target_path, suffix,
+          layer_id,
+          is_balloon_request,
+          arg_map,
+          no_value_arg,
+          cutspec,
+          HTTP_NOT_FOUND);
+      }
+    void TestGoodRequest(
+      request_rec* r,
+      const std::string& source_cmd_or_path,
+      const std::string& target_path,
+      const std::string& suffix,
+      int layer_id,
+      bool is_balloon_request,
+      ArgMap& arg_map,
+      const std::string& no_value_arg,
+      fusion_portableglobe::CutSpec* cutspec,
+      uint64_t expected_data) {
+        TestRequest(
+          r,
+          source_cmd_or_path,
+          target_path, suffix,
+          layer_id,
+          is_balloon_request,
+          arg_map,
+          no_value_arg,
+          cutspec,
+          OK);
+
+        // Check the first few bytes of data written to make sure we're writing
+        // out the correct data.
+        uint64_t written_data;
+        write_data.read(reinterpret_cast<char *>(&written_data), sizeof(written_data));
+        ASSERT_EQ(expected_data, written_data);
+      }
+  private:
+    void TestRequest(
+      request_rec* r,
+      const std::string& source_cmd_or_path,
+      const std::string& target_path,
+      const std::string& suffix,
+      int layer_id,
+      bool is_balloon_request,
+      ArgMap& arg_map,
+      const std::string& no_value_arg,
+      fusion_portableglobe::CutSpec* cutspec,
+      int expected_status) {
+        auto status = service.ProcessPortableRequest(
+          r,
+          source_cmd_or_path,
+          target_path, suffix,
+          layer_id,
+          is_balloon_request,
+          arg_map,
+          no_value_arg,
+          cutspec);
+        ASSERT_EQ(expected_status, status);
+      }
+};
+
+TEST_F(PortableServiceTest, NoUnpacker) {
+  TestBadRequest(r, "command", "wrongtarget", GLOBE, 0, false, args, "", nullptr);
+}
+
+TEST_F(PortableServiceTest, FlatfileQueryNoArgs) {
+  TestBadRequest(r, "flatfile", TARGET, GLOBE, 0, false, args, "query", nullptr);
+}
+
+TEST_F(PortableServiceTest, FlatfileQuery) {
+  args.emplace("blist", "0301324");
+  args.emplace("request", "QTPacket");
+  args.emplace("version", "1");
+  TestGoodRequest(r, "flatfile", TARGET, GLOBE, 0, false, args, "query", nullptr, 0x656c626174726f50);
```

Comment:
Could this parameter be changed to take a string or array of characters as the final parameter?  Or  maybe use a function to convert a string to an uint64_t?

Reply:
Alright, I didn't realize that.  It's a little confusing but I don't have a better solution.